### PR TITLE
Implementation of Lambda Advanced Configuration attribute

### PR DESCRIPTION
**Pull request by Morijarti**
## Motivation
At the moment, AWS support Lambda Advanced configuration, which is caussing issue with snapshot testing since we do not support it in LS.

## Changes
This PR aims to add advanced config to outputs in order to remove various ignores from tests

<!-- The following sections are optional, but can be useful! 

## Testing
Initial test has been created, but removing `$..LoggingConfig`  from lambda related testswill alow  us to test changes after re-generation of snahshots

### CODE REVIEW

**Comment by Morijarti** [comment]
Following tests in `test_lambda_api` are failing
```
tests.aws.services.lambda_.test_lambda_api.TestLambdaFunction.test_create_lambda_exceptions
tests.aws.services.lambda_.test_lambda_api.TestLambdaFunction.test_update_lambda_exceptions
tests.aws.services.lambda_.test_lambda_api.TestLambdaLayer.test_layer_exceptions
```

Do to missing runtime in our error response. AWS shows `ruby3.3` while we don't
AWS response
`Value non-existent-runtime at 'runtime' failed to satisfy constraint: Member must satisfy enum value set: [nodejs20.x, provided.al2023, python3.12, java17, nodejs16.x, dotnet8, python3.10, java11, python3.11, dotnet6, java21, nodejs18.x, provided.al2, ruby3.3, java8.al2, ruby3.2, python3.8, python3.9] or be a valid ARN`

LS response
`Value non-existent-runtime at 'runtime' failed to satisfy constraint: Member must satisfy enum value set: [nodejs20.x, provided.al2023, python3.12, java17, nodejs16.x, dotnet8, python3.10, java11, python3.11, dotnet6, java21, nodejs18.x, provided.al2, java8.al2, ruby3.2, python3.8, python3.9] or be a valid ARN`

**Comment by Morijarti** [comment]
`tests.aws.services.lambda_.test_lambda_api.TestLambdaLayer.test_layer_function_exceptions`
Is failing as it appears that one additional resource is created on AWS side

```
>> match key: list_code_signing_configs
	(~) /CodeSigningConfigs/[0]/CodeSigningConfigArn 'arn:aws:lambda:<region>:111111111111:code-signing-config:<resource:2>' → 'arn:aws:lambda:<region>:111111111111:code-signing-config:<resource:1>' ... (expected → actual)
	(~) /CodeSigningConfigs/[0]/CodeSigningConfigId '<resource:2>' → '<resource:1>' ... (expected → actual)
	(~) /CodeSigningConfigs/[0]/CodeSigningPolicies/UntrustedArtifactOnDeployment 'Enforce' → 'Warn' ... (expected → actual)
	(-) /CodeSigningConfigs/[1] ( {'AllowedPublishers': {'SigningProfileVersionArns': ['arn:aws:signer:<region>:111111111111:/signing-profiles/test']}, 'CodeSigningConfigArn': 'arn:aws:lambda:<region>:111111111111:code-signing-config:<resource:1>', 'CodeSigningConfigId': '<resource:1>', 'CodeSigningPolicies': {'UntrustedArtifactOnDeployment': 'Warn'}, 'Description': 'Testing CodeSigning Config', 'LastModified': 'date'} )
	(+) /CodeSigningConfigs/[1] ( not present )
```

**Comment by Morijarti** [comment]
Following test fails on AWS:
test_cross_account_access

`botocore.exceptions.ClientError: An error occurred (UnrecognizedClientException) when calling the GetLayerVersion operation: The security token included in the request is invalid.`

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 1]
```diff
@@ -78,6 +78,138 @@ def environment_length_bytes(e: dict) -> int:
 
 
 class TestLambdaFunction:
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.py:80]: nit: we could consider grouping these logging config tests under a `TestLoggingConfig` class. 

I usually try ordering the test cases in increasing complexity. Example with some (hierarchical) renaming suggestions:
* TestLoggingConfig.test_advanced_logging_configuration => Cover the basic case first
* TestLoggingConfig.test_log_format_switch
* TestLoggingConfig.test_partial_logging_config_update

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 2]
```diff
@@ -78,6 +78,138 @@ def environment_length_bytes(e: dict) -> int:
...
+
+        aws_client.lambda_.get_waiter("function_updated_v2").wait(FunctionName=function_name)
+
+        received_conf = aws_client.lambda_.get_function_configuration(
+            FunctionName=function_name,
+        )
+        snapshot.match("received_config", received_conf)
+
+    @markers.aws.validated
+    def test_function_advanced_logging_configuration(
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.py:173]: Very clear basic test case 💯

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 3]
```diff
@@ -78,6 +78,138 @@ def environment_length_bytes(e: dict) -> int:
...
+
+        aws_client.lambda_.get_waiter("function_updated_v2").wait(FunctionName=function_name)
+
+        received_conf = aws_client.lambda_.get_function_configuration(
+            FunctionName=function_name,
+        )
+        snapshot.match("received_config_v2", received_conf)
+
+    @markers.aws.validated
+    @pytest.mark.parametrize(
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.py:129]: Kudos for thinking about these edge cases 👏👏👏

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 4]
```diff
@@ -78,6 +78,138 @@ def environment_length_bytes(e: dict) -> int:
...
+        aws_client.lambda_.get_waiter("function_updated_v2").wait(FunctionName=function_name)
+
+        received_conf = aws_client.lambda_.get_function_configuration(
+            FunctionName=function_name,
+        )
+        snapshot.match("received_config", received_conf)
+
     @markers.snapshot.skip_snapshot_verify(
         # The RuntimeVersionArn is currently a hardcoded id and therefore does not reflect the ARN resource update
         # from python3.9 to python3.8 in update_func_conf_response.
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.py:215]: nit: "for different runtime versions" => so we never need to update this comment anymore ;)

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 5]
```diff
@@ -104,7 +236,7 @@ def test_function_lifecycle(self, snapshot, create_lambda_function, lambda_su_ro
 
         update_func_conf_response = aws_client.lambda_.update_function_configuration(
             FunctionName=function_name,
-            Runtime=Runtime.python3_8,
+            Runtime=Runtime.python3_11,
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.py:239]: Thorough update considering the context and realizing not to update here 👏 ✨

File: tests/aws/services/lambda_/test_lambda_api.snapshot.json [hunk 6]
```diff
@@ -5215,26 +5479,18 @@
...
-            },
-            "Description": "Testing CodeSigning Config",
-            "LastModified": "date"
-          }
-        ],
-        "ResponseMetadata": {
-          "HTTPHeaders": {},
-          "HTTPStatusCode": 200
-        }
+        "AllowedPublishers": {
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.snapshot.json:5482]: @dfangl FYI: AWS apparently changed the code signing behavior (see [this backlog item](https://www.notion.so/localstack/Improve-parity-with-Lambda-Code-Signing-Config-6a8432e18105414aad07502023a6f279?pvs=4))

File: tests/aws/services/lambda_/test_lambda_api.snapshot.json [hunk 7]
```diff
@@ -13839,44 +14439,82 @@
           }
         }
       },
-      "update-during-in-progress-update-exc": {
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.snapshot.json:13842]: snapshot diffs are often so confusing 😆

**Review by joe4dev — APPROVED** [review]
Well done @Morijarti with detailed test coverage 👏👏👏 LGTM 👍 

I just added some minor suggestions to improve the naming, then everything is ready to merge 🚀 
For further runtime updates (including ext), it is much better to tackle them in a separate PR.

**Comment by joe4dev** [comment]
@Morijarti I triggered an ext test run against this branch. Please double-check that [this one](https://github.com/localstack/localstack-ext/actions/runs/8722194804) has no failing Lambda tests before merging, thank you!
PS: there's currently a failing websocket test `test_apigatewaymanagementapi` (see [this Slack thread](https://localstack-cloud.slack.com/archives/C02N36URU1G/p1713354222196389))

**Comment by Morijarti** [comment]
@joe4dev I've updated PR with your suggestions. Thank you so much :)

**Comment by Morijarti** [comment]
> @Morijarti I triggered an ext test run against this branch. Please double-check that [this one](https://github.com/localstack/localstack-ext/actions/runs/8722194804) has no failing Lambda tests before merging, thank you! PS: there's currently a failing websocket test `test_apigatewaymanagementapi` (see [this Slack thread](https://localstack-cloud.slack.com/archives/C02N36URU1G/p1713354222196389))

I found few lambda related tests crashing in -ext. I'll create a companion PR for them.

File: tests/aws/services/lambda_/test_lambda_api.snapshot.json [hunk 8]
```diff
@@ -13886,29 +14524,3468 @@
...
+        },
+        "Version": "$LATEST",
+        "ResponseMetadata": {
+          "HTTPHeaders": {},
+          "HTTPStatusCode": 200
+        }
+      }
+    }
+  },
+  "tests/aws/services/lambda_/test_lambda_api.py::TestLambdaFunction::test_function_partial_advanced_logging_configuration_update[partial_config0]": {
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.snapshot.json:14884]: @Morijarti Can you remove these old snapshots after the renaming? Unfortunately, we currently don't have a convenient way to clean up such old snapshot leftovers 🧹

File: tests/aws/services/lambda_/test_lambda_api.py [hunk 9]
```diff
@@ -77,10 +77,144 @@ def environment_length_bytes(e: dict) -> int:
     return string_length_bytes(serialized_environment)
 
 
+class TestLoggingConfig:
+    @markers.aws.validated
+    def test_function_advanced_logging_configuration(
+        self, snapshot, create_lambda_function, lambda_su_role, aws_client
+    ):
+        function_name = f"fn-{short_uid()}"
+        create_response = create_lambda_function(
```
Comment (joe4dev) [inline tests/aws/services/lambda_/test_lambda_api.py:86]: I just noticed that we don't have a test case where we initialize the logging configuration with the function creation. Would it make sense to add that here, given we are testing the default in the partial update below?

File: tests/aws/services/lambda_/test_lambda_api.py (hunk 1 above)
Comment (joe4dev) [reply tests/aws/services/lambda_/test_lambda_api.py]: @Morijarti Do you think the lengthy `advanced_logging_configuration` prefix is necessary now that we have everything grouped in a class?

I think this name in particular is quite long and a bit confusing: `test_function_partial_advanced_logging_configuration_update`
* it's quite long to understand what belongs together: partial config vs. partial update?
* `_function_` doesn't add much (should be clear within Lambda context)
* `_advanced_` doesn't add much (just a confusing name)
* `partial_advanced_logging_configuration` => the name is so long that the `partial_config` part kinda gets lost

File: tests/aws/services/lambda_/test_lambda_api.snapshot.json (hunk 8 above)
Comment (Morijarti) [reply tests/aws/services/lambda_/test_lambda_api.snapshot.json:14884]: Sure thing

File: tests/aws/services/lambda_/test_lambda_api.py (hunk 9 above)
Comment (Morijarti) [reply tests/aws/services/lambda_/test_lambda_api.py:86]: Good catch.
I can add new tests for that.

Comment (joe4dev) [reply tests/aws/services/lambda_/test_lambda_api.py:86]: Feel free to add it here directly. It feels we have already quite extensive API test coverage for a single parameter.

File: tests/aws/services/lambda_/test_lambda_api.snapshot.json (hunk 8 above)
Comment (Morijarti) [reply tests/aws/services/lambda_/test_lambda_api.snapshot.json]: Done

File: tests/aws/services/lambda_/test_lambda_api.py (hunk 9 above)
Comment (Morijarti) [reply tests/aws/services/lambda_/test_lambda_api.py:86]: Good catch, we weren't actually modifying logging config if passed during function creation.
It should be fixed now

File: localstack/services/lambda_/provider.py [hunk 10]
```diff
@@ -869,6 +869,33 @@ def create_function(
                 )
                 # Runtime management controls are not available when providing a custom image
                 runtime_version_config = None
+            if "LoggingConfig" in request:
```
Comment (joe4dev) [inline localstack/services/lambda_/provider.py:872]: glad we caught that one 😃
